File: src/main/java/com/bytetrend/sandbox/java/algo/FindTheTwoElementsWithSmallestDiference.java
package com.bytetrend.sandbox.java.algo;

import java.util.*;
import java.util.function.Supplier;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * https://www.careercup.com/question?id=5071329456291840
 */
public class FindTheTwoElementsWithSmallestDiference {

    public static int[] build(int size) {
        Random r = new Random();
        Supplier<Integer> f = (() -> r.nextInt(200));
        int[] a = new int[size];
        Iterator<Integer> iter = Stream.generate(f).limit(size).collect(Collectors.toCollection(ArrayList::new)).iterator();
        int i = 0;
        while (iter.hasNext()) {
            a[i++] = iter.next();
        }
        return a;
    }

    public static Optional<int[]> findSmallestDiff(int[] a) {
        int[] r = new int[]{a[0], a[1]};
        int i = 2;
        while (i < a.length) {
            //If smallest than previous
            if ((r[1] - r[0]) > (a[i] - a[i - 1])) {
                r[0] = a[i - 1];
                r[1] = a[i];
            }
            i++;
        }
        return Optional.of(r);
    }

    public static Optional<int[]> find(int[] a) {
        if (a.length < 2)
            return Optional.empty();
        else {
            Arrays.sort(a);
            return findSmallestDiff(a);
        }
    }

    public static void main(String[] args) {
        int[] arr = build(10);
        Arrays.sort(arr);
        System.out.println(String.format("%s %s", Arrays.toString(arr), Arrays.toString(find(arr).orElseGet(() -> new int[]{}))));
    }
}

File: src/main/scala/com/bytetrend/sandbox/scala/algo/FindTheTwoElementsWithSmallestDiference.scala
package com.bytetrend.sandbox.scala.algo

import scala.util.Random

/**
  * https://www.careercup.com/question?id=5071329456291840
  */
object FindTheTwoElementsWithSmallestDiference extends App {

  var arr = Seq.fill(10)(Random.nextInt(100)).toArray

  def find(a: Seq[Int]): Option[(Int, Int)] = {
    a.length match {
      case x if (x < 2) => None
      case _ => Some(findSmallestDiff(a.sorted))
    }
  }

  def findSmallestDiff(a: Seq[Int]): (Int, Int) = {
    var sml = (a.head, a.tail.head)
    a.tail.tail.foldLeft(sml._2-sml._1,a.tail.head) { case ((diff,last), n) => {
      if (diff < (n - last))
        (diff, n)
      else {
        sml = (last,n)
        (n - last, n)
      }
    }
    }
    sml
  }

  println(s"for ${arr.sorted.mkString(",")} ${find(arr)}")
  arr = Seq.fill(6)(Random.nextInt(100)).toArray
  println(s"for ${arr.sorted.mkString(",")} ${find(arr)}")
  arr = Seq.fill(2)(Random.nextInt(100)).toArray
  println(s"for ${arr.sorted.mkString(",")} ${find(arr)}")
  arr = Seq.fill(1)(Random.nextInt(100)).toArray
  println(s"for ${arr.sorted.mkString(",")} ${find(arr)}")

}
